source/building-your-own-javascript-components/index.html.md.erb

```diff
@@ -0,0 +1,283 @@
+
+You can work around this issue by defining the ‘$root’ property in your component class yourself.
+
+```js
+class MyComponent extends Component {
+  static moduleName = ‘app-my-component’
+
+  // Defining the property here makes TypeScript aware of its existence
+  $root
+
+  constructor($root) {
+    super($root)
+
+    // Run component specific initialisation here, for example:
+    this.$root.addEventListener(‘click’, this.handleClick.bind(this))
+  }
+
+  handleClick(event) {
+    this.$root.classList.add(‘app-my-component--clicked’)
+  }
+}
+```
+
+If you’re interested in better TypeScript support for GOV.UK Frontend, please make us aware on the GitHub issue for [exporting type declarations](https://github.com/alphagov/govuk-frontend/issues/2835). 
+
+
+### Customising how support is checked
+
+The `Component` class will check that GOV.UK Frontend is supported during initialisation. However, your component may have different requirements for running properly. For example, it may require specific JavaScript APIs that are not supported by all browsers.
+
+You can redefine the static `checkSupport` method so the component throws an error if these requirements are not met. We recommend adding [the component’s `moduleName`](#defining-the-modulename-property) to help error messages identify which component they relate to without having to search their stack trace.
+
+```js
+class MyComponent extends Component {
+  static moduleName = ‘app-my-component’
+
+  static checkSupport() {
+    // Run the same checks as the `Component` class
+    Component.checkSupport()
+
+    // Check for support of the native clipboard API
+    if (!navigator.clipboard) {
+      throw new Error(‘Clipboard API not supported’)	
+    }
+  }
+}
+```
+## Using `createAll` with your components
+
+Use the `createAll` function to initialise your components [the same way GOV.UK Frontend components are initialised](../import-javascript/#import-individual-components). The function will look up all the HTML elements on which a given component should be initialised and initialise a component for each of them.
+
+```js
+import {createAll} from ‘govuk-frontend’
+import {ProjectComponent, OtherProjectComponent} from ‘./project-components.js’
+
+createAll(ProjectComponent)
+// You can provide a config for components that use them
+createAll(OtherProjectComponent, {
+  anOption: ‘itsValue’
+})
+```
+
+The `createAll` function also lets you:
+
+- [only search a specific part of the page for elements](#initialise-only-on-part-of-a-page) (for example, after adding new content with JavaScript)
+- [respond to errors during components initialisation](#handling-initialisation-errors) (for example, sending them to an error reporting service)
+
+To avoid reimplementing features shared across components, we recommend that your components extend [our `Component` class](#using-the-component-class). However, this is not a requirement, and all a component needs to be used with `createAll` is:
```

Comment:
Tried to make the distinction between our recommendation and the bare requirement a bit clearer.

```suggestion
To avoid reimplementing features shared across components, we recommend that your components extend [our `Component` class](#using-the-component-class). However, you can choose not to do this. The only requirements for a component to be used with `createAll` are:
```